Jetpack: redirect to the Jetpack page on multisites after activation

sergeymitr opened the pull request:
Related to #35449
Fixes #35557

## Proposed changes:
When Jetpack is activated, we redirect to My Jetpack, which is unavailable on multisites.
This PR will redirect users on multisites to the Jetpack page.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
p1707442008656029-slack-CA17PSW87

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
1. Setup a non-multisite site.
2. Go to "Plugins" and deactivate Jetpack and all standalone plugins.
3. Activate Jetpack. Confirm you got redirected to the "My Jetpack" page.
4. If you're using a new site without any standalone plugins installed, "My Jetpack" should display the "Welcome to Jetpack!" banner.
5. Setup a multisite, go to "Plugins" and deactivate Jetpack and all standalone plugins.
6. Network-activate Jetpack, confirm you're redirected to `/network/admin.php?page=jetpack`.
7. Network-deactivate Jetpack.
8. Activate Jetpack on a subsite, confirm you're redirected to the Jetpack's connect screen.

sergeymitr committed:
Jetpack: redirect to the Jetpack page on multisites after activation.

oskosk commented:
Maybe, if the plugin is being network-activated, we should redirect to 
`wp-admin/network/admin.php?page=jetpack.`

<img width="1207" alt="image" src="https://github.com/Automattic/jetpack/assets/746152/086dc3d9-3884-4d6d-9864-768198e35321">


Right now, in this PR, we are checking for multisite and redirecting to 

`/wp-admin/admin.php?page=jetpack`, which at the same time, is the next destination if the user clicks the Set up Jetpack button in the network-wide Jetpack screen. 

The behavior in this PR could work, but this behavior implies that, by activating Jetpack, the user gets moved from the network context to the main site context. 

What do you think?

jeherve reviewed (commented):
This tested well for me. 

Maybe another way to work around it would be to call `Automattic\Jetpack\My_Jetpack\Initializer` here, and use the `should_initialize` method here instead of a hardcoded multisite check. This way when we eventually make My Jetpack available to Multisites, your logic would get automatically updated.

jeherve commented:
> The behavior in this PR could work, but this behavior implies that, by activating Jetpack, the user gets moved from the network context to the main site context.

That's a good point, good catch.

We may need to differentiate between network-activations and site activations:

- When network-activating, I would expect to stay within the network admin. As a superadmin, I will not necessarily be the one to manage the connection to WordPress.com for each site. That's something I will want to decide on a site per site basis. The network admin screen is good for that.
- When site activating, I would expect to be directed to the Jetpack plugin for that site.

sergeymitr committed:
Merge branch 'trunk' into fix/jetpack-activation-redirect-on-multisites

sergeymitr committed:
Adjust redirect pages and priorities.

sergeymitr commented:
Hi @oskosk, @jeherve
Thanks for testing and the ideas.
I implemented the changes you suggested, and also added an exception for WoA sites as they also don't have My Jetpack.
Please give it another look when you get a few minutes.

Commit:
Merge branch 'trunk' into fix/jetpack-activation-redirect-on-multisites

jeherve commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -6896,7 +6896,17 @@ private function plugin_post_activation() {
 			delete_transient( 'activated_jetpack' );
 
-			wp_safe_redirect( static::admin_url( 'page=my-jetpack' ) );
+			if ( ( new Host() )->is_woa_site() ) {
+				$redirect_url = static::admin_url( 'page=jetpack' );
+			} elseif ( is_network_admin() ) {
```
Do we need this WoA use-case? Since Jetpack is a managed plugin on Atomic, are there ever situations when the plugin is not enabled?

sergeymitr replied:
I believe you need Jetpack Beta to deactivate Jetpack on WoA, but it's still possible.
So we could get by without it, but I believe having it there is better so site owners wouldn't run into the error after some occasional experiments.

jeherve reviewed (approved):
This tests well, and should be a good fix until #35558.

I only have a small remark.

samiff commented:
cherry-picked to `jetpack/branch-13.1.2` in 7cd254f2666b41161084905292b0c223c10b8899
